sdk/pinocchio/src/cpi.rs

```
@@ -50,34 +68,44 @@ pub fn invoke_signed<const ACCOUNTS: usize>(
-        let account_meta = &instruction.accounts[index];
-
-        if account_info.key() != account_meta.pubkey {
-            return Err(ProgramError::InvalidArgument);
-        }
-
-        let state = if account_meta.is_writable {
-            BorrowState::Borrowed
-        } else {
-            BorrowState::MutablyBorrowed
-        };
-
-        if account_info.is_borrowed(state) {
-            return Err(ProgramError::AccountBorrowFailed);
-        }
-
-        accounts[index].write(Account::from(account_infos[index]));
-    }
-
+    account_infos
+        .iter()
+        .zip(instruction.accounts.iter())
+        .zip(accounts.iter_mut())
+        .try_for_each(|((account_info, account_meta), accounts)| {
+            // In order to check whether the borrow state is compatible
+            // with the invocation, we need to check that we have the
+            // correct account info and meta pair.
+            if account_info.key() != account_meta.pubkey {
+                return Err(ProgramError::InvalidArgument);
+            }
+
+            // Determines the borrow state that would be invalid according
+            // to their mutability on the instruction.
+            let invalid_state = if account_meta.is_writable {
+                // If the account is required to be writable, it cannot
+                // be currently borrowed.
+                BorrowState::Borrowed
+            } else {
+                // If the account is required to be read-only, it cannot
+                // be currently mutably borrowed.
+                BorrowState::MutablyBorrowed
+            };
+
+            if account_info.is_borrowed(invalid_state) {
+                return Err(ProgramError::AccountBorrowFailed);
+            }
+
+            accounts.write(Account::from(*account_info));
+
+            Ok(())
+        })?;
+
+    // SAFETY: At this point it is guaranteed that all account infos are
+    // borrowable according to their mutability on the instruction.
```

I do not know if this matters, but if `instruction.accounts` is shorter than `ACCOUNTS`, not all `ACCOUNTS` elements in the `accounts` array will be initialized.

---

Yes, we should avoid that – I just pushed this change + a few small tweaks to the imports.

---

I thought about it a bit more, and I wonder if `min` is really the right behavior here.
`instruction` argument describes the invocation that the caller of the `invoke_signed()` or the `slice_invoke_signed()` is requesting.
When they specify a certain value for the `accounts` field, they probably expect the invocation to contain exactly this many accounts.
So, it seems that we should rather panic or return `Err(ProgramError::InvalidArgument)` when the actual number of accounts turns out to be different from `instruction.accounts.len()`.

All the other errors are reported as `Err(ProgramError::Something)`, even a `account_infos.len() > MAX_CPI_ACCOUNTS` violation.  So, maybe this case should also be an `Err()` result.
At the same time, it seems that the `account_infos.len() > MAX_CPI_ACCOUNTS` and the insufficient number of accounts are API violations, which are better handled by a panic, rather than an error.

---

The idea was to simplify the invoke and let the syscall fail in cases where there are accounts missing since the syscall will validate the input as well without consuming more CUs of the program – I updated the description of the PR to reflect this. For some reason, adding a simple `if` statement to test whether the length mismatches adds `100` CUs. 🤷‍♂️

But perhaps we won't be able to avoid it since there is an edge case that we won't be able to validate. Assume an example that we require `5` accounts on the instruction (the last one required to be writable) but we pass only `4` account infos. We can only validate the borrow of the  `4` account infos, which corresponds to the first `4` account metas of the instruction. Now imagine that the `5`th one is a duplicated account that matches one of the given account infos. The syscall is not going to fail since all accounts have been passed in, but the problem is that we did not validate that it is safe to borrow account `5` as writable – this is a problem!

cc: @LucasSte to see if he can help us understand why adding the `if` statement is so costly in this case.

---

Did a quick comparison using the same instruction with 2 accounts:
* `invoke`: `1315` CUs
* `slice_invoke`: `1468` CUs (+153)

Apart from statically allocating a large `MaybeUninit` array and the `if` statement validating the accounts length, they are very similar functions.

---

> to see if he can help us understand why adding the if statement is so costly in this case.

I read the conversation, but I didn't follow which if statement you'd like to include here. Is it the one in the PR description?

---

> > to see if he can help us understand why adding the if statement is so costly in this case.
> 
> I read the conversation, but I didn't follow which if statement you'd like to include here. Is it the one in the PR description?

Yes, something like:
```rust
if instruction.accounts.len() != ACCOUNTS {
    return Err(ProgramError::NotEnoughAccountKeys);
}
```
At the very start of the `invoke_signed` function.

---

Even though there is the `#[inline]` attribute on top of the `invoke_signed` function, when you add the if, the function is not inlined anymore. The hint does not guarantee the function will be inlined, and I believe the extra length makes inlining not worth as the penalty for code size would be too large.

---

Oh, you are right, nice! So what would be the best alternative? If we add `#[inline(always)]` to `invoke_signed` then it can save `100` CUs. The drawback would be if you try to call it with a large `ACCOUNTS` value, you might end up with no stack. Would be acceptable a workaround like [this](https://nnethercote.github.io/perf-book/inlining.html#harder-cases)?

---

> Would be acceptable a workaround like [this](https://nnethercote.github.io/perf-book/inlining.html#harder-cases)?

Yes, I believe that is a reasonable workaround.